### DISCUSSION TITLE
saving learning queries as a file

### DISCUSSION THREAD

**Discussion by Szpilman2** [UPPAALModelChecker/UPPAAL-Meta#238] [Q&A] [answered]
Dear UPPAAL team,

Hello! I'm interested in saving my learning strategy as a file, but it seems that the saved file doesn't adhere to the strategy file template. Could you please guide me on how to properly save my learning strategy?

Strategy file for ```strategy teststra = control: A<> m.sofLoc2``` is : 
[ControllerStrategy.json](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/13612230/ControllerStrategy.json)

Strategy file for ```strategy myStrategy = minE(energy) [<=200] : <> m.sofLoc2 under teststra``` is: 
[LearningStrategy.json](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/13613420/LearningStrategy.json)

as you can see the ControllerStrategy.json file has not structure of ControllerStrategy.json. 

Thank you.

**Comment by thorulf4** [comment]
I think you accidentally uploaded the same file twice as the 2 attachments are the same, but I'll assume that was just a mistake.

But it seems to be working as intended, as we don't expect all strategies to be saved in the same format. Broadly speaking we have two categories of strategies synthesized strategies and learning strategies. These two strategies are stored in different formats I assume that is where your confusion comes from. 
Both of these strategy types can be saved and loaded using the saveStrategy or loadStrategy [syntax](https://docs.uppaal.org/language-reference/query-syntax/strategy_queries/)
#### Synthesized strategies
So symbolic queries like:
`control: A<> ...`
`control: A[] ...`
Synthesized strategies are saved in a custom non Json format.

#### Learning strategies
So queries like:
`minE(cost) ... `
`maxE(reward) ... `
Learning strategies are stored in json

**Reply by Szpilman2** [reply]
Thank you so much. My apologies for the confusion; the files in question are: (I also updated my original question)
[ControllerStrategy.json](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/13613384/ControllerStrategy.json)
[LearningStrategy.json](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/13613386/LearningStrategy.json)

also my uppaal model is:
![model](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/d9d15298-8f44-4054-90c1-7752b986a7dd)


I believe my question might have been a bit unclear. My concern is that the content in "LearningStrategy.json" doesn't seem to constitute a strategy when compared to "ControllerStrategy.json." 

in other words I think the content of LearningStrategy.json is not a strategy. and structure of files (they are json formatted or not) is not my concern.  

Thank you very much.

**Reply by thorulf4** [reply]
Ah thanks for clarifying your question! But your concern is correct, your `LearningStrategy.json` has not learnt anything. Which happens because the `error` location was never reached during learning. As you're using an exact guard, which means your error state will only be reached when a delay of exactly 5 is chosen which is almost impossibly unlikely

To fix this I would rewrite the error bound using a guard and a urgent channel, which makes 5 an upper bound where a decision must be made (Red area). That way we get to use rate of exponential which has better semantics for your use case. To handle the case where a delay less than 1 I add a Delay action (Blue area). 
```C++
// Place global declarations here.
urgent broadcast chan ASAP;
clock x, energy;
```
![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/10059450/7838eb51-9e6c-4417-a6a5-cd9c5fa78bf3)
Also your liveness strategy may also prune all but one action which will also make training produce empty strategies.

**Reply by Szpilman2** [reply]
Thank you very much for your reply. I have a couple of inquiries:

1. Could you explain the purpose of incorporating the "ASAP" channel in this specific context? Essentially, why is this channel included in the transition? Additionally where should I put "ASAP?" in the model?

2. While the invariant in the initial location (sofLoc1) is 5 (x <= 5), I'm having trouble discerning the precise delay value in the blue area. Based on your previous statement, it appears to be 1, but could you clarify this (that how can i change this to five without getting empty strategy, in other words I would like to have x<=5 as invariant and the upper bound for invariant is fixed.)?
3. Additionally, I believe the energy should be represented by a "hybrid clock" in this scenario. Is my understanding correct?

Thanks a lot.

**Reply by thorulf4** [reply, marked as answer]
1) ASAP is a channel marked as an `urgent` which means we cannot delay while the transition is available. Which means we can't delay past 5 as the edge becomes available at `x>=5`. This will effectively impose a `x<=5` invariant ASAP is a channel declared with `urgent broadcast chan ASAP;` and it is used in the synchronization field on the edge. [Documenation on urgent channels](https://docs.uppaal.org/language-reference/system-description/templates/edges/#synchronisations)
2) This model does not have a specific delay, it uses [Rate of Exponential](https://docs.uppaal.org/language-reference/system-description/templates/locations/) which picks a delay randomly with an exponential distribution where small delays are more likely. This does actually break the `x<=5` invariant. If you prefer you can keep the invariant if you're okay with having constrained delays. In this example I use the `clock y` to force the delay to be 1 time unit 
![image](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/10059450/97b00e1a-835b-4536-b84d-18d61394a3b6)
3) I'm not entirely sure if you need hybrid clock here. I think It depends what exactly you need to use the model for

**Reply by Szpilman2** [reply]
Thank you for taking the time to address my question; I highly value your insights and appreciate your assistance.

I think my question was not clear enough so I would like to explain it again. Regarding my scenario, I have two automata, labeled m and n. Automaton m comprises two states, sofLoc1 and sofLoc2. In sofLoc1, we can wait for a maximum of 5 time units (as an invariant). Upon the clock value surpassing 1, a transition to sofLoc2 (the goal state) can be taken. automata m and n are synchronized through ```use``` channel. Additionally, there is an automaton to monitor the energy consumption, where automaton n represents specific hardware with each state consuming a certain amount of energy (3 units in this case). in other words ```energy = 3 * t``` where t is the time parameter.
![289057011-d9d15298-8f44-4054-90c1-7752b986a7dd](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/042a7ba4-844c-41b8-b8d9-bd13c8e3970c)

While I successfully modeled this scenario, attempting to verify the property `minE(energy) [<=10] : <> m.sofLoc2` in UPPAAL resulted in a "time lock error." (visit [this](https://github.com/orgs/UPPAALModelChecker/discussions/237)) In search of a solution, I came across a [fix](https://github.com/orgs/UPPAALModelChecker/discussions/106) suggesting the introduction of an error state. I implemented this fix, specifying that when the clock value reaches 5, it should transition to the error state with a controllable transition with guard ```x==5```. However, this led to another time lock error. Subsequently, I changed the controllable transition to uncontrollable, resolving the time lock error but resulting in an empty strategy when attempting to find a strategy for `minE(energy) [<=10] : <> m.sofLoc2`.
![289020838-6ab52507-0cde-4e0b-95e9-71735cd60fe5](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/2d05f078-333f-4f3c-a130-03c38b967ae2)
![289020910-6afb06e1-54ff-46bf-bdd2-6604031cb064](https://github.com/UPPAALModelChecker/UPPAAL-Meta/assets/35931007/b3c4400b-d851-441d-93b0-83c8dcceb0c9)


Now, I am seeking guidance on how to address this issue (how to modify automata to get valid strategy). I think modifying the guard from `x==5` to `x>=5` or removing the `x==5` condition, might resolve the problem and yield a valid strategy (non-empty). I've reviewed your previous responses, and while I understand that adding exponential time delay may not be applicable in my scenario, Could you please provide assistance in resolving this matter?

**Reply by thorulf4** [reply]
Okay, I hope I understand the question better now. I think you get an empty strategy because there is only 1 possible action for the strategy to choose i.e there is only one controllable edge. If you want the strategy to decide if to take the softLoc1 --> softLoc2 transition or delaying you need to make a controllable transition specifically for delaying.
